site/generator/publisher.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+import dataclasses
+import io
+import os
+import re
+import shutil
+import sys
+
+from scss.compiler import compile_file
+import click
+
+from generator import models
+from generator import env
+
+
+class Publisher:
+    """Class for writing the AIP site out to disk."""
+
+    def __init__(self, output_dir, *, log_to_stdout=False):
+        self.output_dir = output_dir.rstrip(os.path.sep)
+        self.log_to_stdout = log_to_stdout
+        self.site = models.Site.load()
+
+    def log(self, message: str, *, err: bool = False):
+        """Conditionally log a message."""
+        dest = sys.stderr if err else sys.stdout
+        if self.log_to_stdout:
+            print(f'{message}', file=dest)
+
+    def publish_site(self):
+        """Publish the full site."""
+        for page in self.site.pages.values():
+            self.publish_doc(page)
+        for news in self.site.news.values():
+            self.publish_doc(news)
+        for scope in self.site.scopes.values():
+            self.publish_doc(scope)
+        for aip in self.site.aips.values():
+            self.publish_doc(aip)
+        self.publish_template('index.html.j2', 'index.html')
+        self.publish_template('search.html.j2', 'search.html')
+        self.publish_static()
+        self.publish_template(
+            'search.js.j2',
+            'static/js/search/tipuesearch_content.js',
+        )
+        self.publish_css()
+
+    def publish_doc(self, doc: Union[models.AIP, models.Scope, models.Page]):
+        """Publish a single AIP document to disk."""
+        # Determine the path and make sure the directory exists.
+        path = f'{self.output_dir}{doc.relative_uri}.html'
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+
+        # Write the document's file to disk.
+        with io.open(path, 'w') as f:
+            f.write(doc.render())
+        self.log(f'Successfully wrote {path}.')
+
+        # If this document has any redirects (only AIPs do as of this writing),
+        # write out the redirect files to disk also.
```

Comment:
to disk? ew...

I'm guessing the reasoning is that this is more performant at runtime, perhaps? Although I don't think storing a redirect map and checking it would be too costly, only only checking it if a file doesn't exist maybe?

Reply:
This is the publish step; there is no redirect map to check. (Worth noting, this is exactly duplicating how Jekyll works.)